tests/worker/test_workflow.py

```diff
@@ -6146,3 +6162,45 @@ async def test_workflow_run_sees_workflow_init(client: Client):
             task_queue=worker.task_queue,
         )
         assert workflow_result == "hello, world"
+
+
+@workflow.defn
+class UpdateCancellationWorkflow:
+    def __init__(self) -> None:
+        self.non_terminating_operation_has_started = False
+
+    @workflow.run
+    async def run(self) -> NoReturn:
+        await asyncio.Future()
+
+    @workflow.update(unfinished_policy=workflow.HandlerUnfinishedPolicy.ABANDON)
+    async def non_terminating_update(self) -> NoReturn:
+        self.non_terminating_operation_has_started = True
+        await asyncio.Future()
+
+    @workflow.update
+    async def wait_until_non_terminating_operation_has_started(self) -> None:
+        await workflow.wait_condition(
+            lambda: self.non_terminating_operation_has_started
+        )
+
+
+async def test_update_cancellation(client: Client):
```

Comment:
May be worth having a test for doing a cleanup step on cancellation (e.g. in a `finally`). Can do this in the update _or_ the workflow. This is common in Temporal code and I fear cleanup code in an update with issue a warning.

Reply:
I see, good point. So to restate what you're saying:

> If a workflow cancellation request comes in, and a handler is unfinished, the workflow author should be able to allow the workflow to be cancelled, and to have it gracefully interrupt the handler, and not to be criticized by any warning.

That is true today: we don't cancel signals or updates on workflow cancellation, but a user could write a workflow that _does_, and they would escape the warning, because the handler would not be alive at check time.

But my PR breaks it, by attempting to make the check at workflow cancellation time.

The reason my PR does that is that it wanted to continue to warn the workflow author when workflow cancellation _causes_ an update cancellation. And the reason it wants to do that is to maintain consistency with signals: when the workflow is canceled, if we're going to warn on unfinished signal handlers (which we do) then we should warn on unfinished update handlers also.

I think that these desires are mutually incompatible. We have to allow a user to clean up their update gracefully without the warning, as you point out. I see two solutions:

1. to change things so that we no longer issue unfinished handler warnings for either signal or update, in the case of workflow cancellation.

2. to automatically cancel signals also (with a workflow logic flag)

There is a strong incentive (caller UX) for automatically canceling update handlers on workflow cancellation. We have never automatically cancelled signal handlers in the Python SDK, and doing so would require a flag as it may be backwards incompatible with replay of existing histories. The inconsistency in cancelling update handlers but not signal handlers would only be a problem if there is a potential for observable side effects.

I'll carry on thinking about it but for now I've updated the PR to stop issuing the unfinished warning on workflow cancellation, for both signal and update.

Reply:
You're suggesting that we continue to do the check late, and warn on workflow cancellation.

We can't do that, because in the common case we will then warn on unfinished signals (since these are not automatically cancelled, and hence the handlers will still be alive) but not warn on unfinished updates (since these are automatically cancelled, and hence the handlers will be dead).

(I did try to explain that in the comment above https://github.com/temporalio/sdk-python/pull/653#discussion_r1779214147, but it is turning out to be a slightly tricky subject! And my comment isn't very clear).

Reply:
To confirm from our off-PR discussion, we are still warning on unfinished handler regardless of reason (i.e. cancellation)?

Reply:
Yes, still thinking about the design here. I've put the PR into draft mode for now.